// @isnotdefined/stylelint-plugin/index.js
import stylelint from 'stylelint';
import { parse } from 'postcss-values-parser';

const wording = {
    expected: 'Expected',
    unexpected: 'Unexpected',
    atRule: 'at-rule',
    comment: 'comment',
    property: 'property',
    value: 'value',
    unit: 'unit',
    toBe: 'to be'
};

const defaultOptions$2 = {
    commands: [
        'stylelint-disable',
        'stylelint-disable-line',
        'stylelint-disable-next-line'
    ]
};

const { utils: utils$2, createPlugin: createPlugin$2 } = stylelint;
const ruleName$2 = '@isnotdefined/no-disable';
function validateOptions$2(result, options) {
    return utils$2.validateOptions(result, ruleName$2, {
        actual: options,
        // @ts-ignore
        possible: {
            commands: value => value
        }
    });
}
function rule$2(primaryOptions) {
    const options = Object.assign(Object.assign({}, defaultOptions$2), primaryOptions);
    return (root, result) => {
        if (validateOptions$2(result, options)) {
            root.walkComments(comment => {
                if (options.commands.includes(comment.text.split(' ').at(0))) {
                    utils$2.report({
                        message: wording.unexpected + ' "' + comment.text + '" ' + wording.comment,
                        node: comment,
                        result,
                        ruleName: ruleName$2,
                        word: comment.text
                    });
                }
            });
        }
    };
}
var noDisableRule = createPlugin$2(ruleName$2, rule$2);

const defaultOptions$1 = {
    atRules: [
        {
            name: {
                search: 'apply'
            }
        },
        {
            name: {
                search: 'document'
            }
        },
        {
            name: {
                search: 'viewport'
            }
        }
    ],
    properties: [
        {
            name: {
                search: 'box-align'
            }
        },
        {
            name: {
                search: 'box-direction',
                replace: 'flex-direction'
            }
        },
        {
            name: {
                search: 'box-flex'
            }
        },
        {
            name: {
                search: 'box-flex-group'
            }
        },
        {
            name: {
                search: 'box-lines'
            }
        },
        {
            name: {
                search: 'box-ordinal-group',
                replace: 'order'
            }
        },
        {
            name: {
                search: 'box-orient'
            }
        },
        {
            name: {
                search: 'box-pack'
            }
        },
        {
            name: {
                search: 'box-sizing'
            },
            value: {
                search: 'padding-box'
            }
        },
        {
            name: {
                search: 'clip',
                replace: 'clip-path'
            }
        },
        {
            name: {
                search: 'grid-column-gap',
                replace: 'column-gap'
            }
        },
        {
            name: {
                search: 'grid-gap',
                replace: 'gap'
            }
        },
        {
            name: {
                search: 'grid-row-gap',
                replace: 'row-gap'
            }
        },
        {
            name: {
                search: 'image-orientation'
            },
            value: {
                search: 'flip'
            }
        },
        {
            name: {
                search: 'image-rendering'
            },
            value: {
                search: 'optimizeQuality',
                replace: 'smooth'
            }
        },
        {
            name: {
                search: 'image-rendering'
            },
            value: {
                search: 'optimizeSpeed',
                replace: 'pixelated'
            }
        },
        {
            name: {
                search: 'ime-mode'
            }
        },
        {
            name: {
                search: 'page-break-after',
                replace: 'break-after'
            }
        },
        {
            name: {
                search: 'page-break-before',
                replace: 'break-before'
            }
        },
        {
            name: {
                search: 'scroll-snap-coordinate'
            }
        },
        {
            name: {
                search: 'scroll-snap-destination'
            }
        },
        {
            name: {
                search: 'scroll-snap-points-x'
            }
        },
        {
            name: {
                search: 'scroll-snap-points-y'
            }
        },
        {
            name: {
                search: 'scroll-snap-type-x'
            }
        },
        {
            name: {
                search: 'scroll-snap-type-y'
            }
        },
        {
            name: {
                search: 'text-decoration'
            },
            value: {
                search: 'blink'
            }
        },
        {
            name: {
                search: 'text-decoration-line'
            },
            value: {
                search: 'blink'
            }
        },
        {
            name: {
                search: 'user-modify'
            }
        },
        {
            name: {
                search: 'word-break',
                replace: 'overflow-wrap'
            },
            value: {
                search: 'break-word'
            }
        },
        {
            name: {
                search: 'word-wrap',
                replace: 'overflow-wrap'
            }
        }
    ]
};

const { utils: utils$1, createPlugin: createPlugin$1 } = stylelint;
const ruleName$1 = '@isnotdefined/no-obsolete';
function validateOptions$1(result, options) {
    return stylelint.utils.validateOptions(result, ruleName$1, {
        actual: options,
        // @ts-ignore
        possible: {
            atRules: value => value,
            properties: value => value
        }
    });
}
function rule$1(primaryOptions, secondaryOptions, context) {
    const options = Object.assign(Object.assign({}, defaultOptions$1), primaryOptions);
    return (root, result) => {
        var _a, _b;
        if (validateOptions$1(result, options)) {
            (_a = options.atRules) === null || _a === void 0 ? void 0 : _a.map(atRule => {
                var _a;
                root.walkAtRules((_a = atRule === null || atRule === void 0 ? void 0 : atRule.name) === null || _a === void 0 ? void 0 : _a.search, decl => {
                    const { name } = atRule;
                    if (name === null || name === void 0 ? void 0 : name.replace) {
                        utils$1.report({
                            message: wording.expected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.atRule + ' ' + wording.toBe + ' "' + (name === null || name === void 0 ? void 0 : name.replace) + '"',
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.name
                        });
                        if (context.fix) {
                            decl.name = name === null || name === void 0 ? void 0 : name.replace;
                        }
                    }
                    else {
                        utils$1.report({
                            message: wording.unexpected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.atRule,
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.name
                        });
                    }
                });
            });
            (_b = options.properties) === null || _b === void 0 ? void 0 : _b.map(property => {
                var _a;
                root.walkDecls((_a = property === null || property === void 0 ? void 0 : property.name) === null || _a === void 0 ? void 0 : _a.search, decl => {
                    const { name, value } = property;
                    const valueMatch = (value === null || value === void 0 ? void 0 : value.search) === decl.value;
                    if (!(value === null || value === void 0 ? void 0 : value.search) && (name === null || name === void 0 ? void 0 : name.replace) || valueMatch && (name === null || name === void 0 ? void 0 : name.replace)) {
                        utils$1.report({
                            message: wording.expected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.property + ' ' + wording.toBe + ' "' + (name === null || name === void 0 ? void 0 : name.replace) + '"',
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.prop
                        });
                        if (context.fix) {
                            decl.prop = name === null || name === void 0 ? void 0 : name.replace;
                        }
                    }
                    else if (!(value === null || value === void 0 ? void 0 : value.search)) {
                        utils$1.report({
                            message: wording.unexpected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.property,
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.prop
                        });
                    }
                    else if (valueMatch && (value === null || value === void 0 ? void 0 : value.replace)) {
                        utils$1.report({
                            message: wording.expected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.value + ' "' + (value === null || value === void 0 ? void 0 : value.search) + '" ' + wording.toBe + ' "' + (value === null || value === void 0 ? void 0 : value.replace) + '"',
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.value
                        });
                        if (context.fix) {
                            decl.value = value === null || value === void 0 ? void 0 : value.replace;
                        }
                    }
                    else if (valueMatch && !(value === null || value === void 0 ? void 0 : value.replace)) {
                        utils$1.report({
                            message: wording.unexpected + ' "' + (name === null || name === void 0 ? void 0 : name.search) + '" ' + wording.value + ' "' + (value === null || value === void 0 ? void 0 : value.search) + '"',
                            node: decl,
                            result,
                            ruleName: ruleName$1,
                            word: decl.value
                        });
                    }
                });
            });
        }
    };
}
var noObsoleteRule = createPlugin$1(ruleName$1, rule$1);

const defaultOptions = {
    step: 0.125,
    units: [
        'ch',
        'em',
        'ex',
        'rem',
        'vh',
        'vw',
        'vmin',
        'vmax'
    ],
    atRules: [
        'media'
    ],
    properties: [
        'border',
        'border-bottom',
        'border-bottom-left-radius',
        'border-bottom-right-radius',
        'border-bottom-width',
        'border-left',
        'border-left-width',
        'border-radius',
        'border-right',
        'border-right-width',
        'border-top',
        'border-top-left-radius',
        'border-top-right-radius',
        'border-top-width',
        'bottom',
        'box-shadow',
        'columns',
        'column-gap',
        'column-width',
        'flex-basis',
        'font',
        'font-size',
        'gap',
        'height',
        'inset',
        'left',
        'letter-spacing',
        'line-height',
        'margin',
        'margin-bottom',
        'margin-left',
        'margin-right',
        'margin-top',
        'max-height',
        'max-width',
        'min-height',
        'min-width',
        'outline',
        'outline-offset',
        'outline-width',
        'padding',
        'padding-bottom',
        'padding-left',
        'padding-right',
        'padding-top',
        'right',
        'row-gap',
        'text-indent',
        'top',
        'width'
    ]
};

const { utils, createPlugin } = stylelint;
const ruleName = '@isnotdefined/unit-step';
function validateOptions(result, options) {
    return utils.validateOptions(result, ruleName, {
        actual: options,
        // @ts-ignore
        possible: {
            step: value => value > 0,
            units: value => defaultOptions.units.includes(value),
            atRules: value => defaultOptions.atRules.includes(value),
            properties: value => defaultOptions.properties.includes(value)
        }
    });
}
function rule(primaryOptions, secondaryOptions, context) {
    const options = Object.assign(Object.assign({}, defaultOptions), primaryOptions);
    return (root, result) => {
        var _a, _b;
        if (validateOptions(result, options)) {
            (_a = options.atRules) === null || _a === void 0 ? void 0 : _a.map(atRule => {
                root.walkAtRules(atRule, decl => {
                    const params = parse(decl.params);
                    params.walkNumerics(param => {
                        if (options.units.includes(param.unit) && Number(param.value) % options.step > 0) {
                            const valueFixed = Math.round(Number(param.value) / options.step) * options.step;
                            utils.report({
                                message: wording.expected + ' "' + atRule + '" ' + wording.unit + ' "' + param.value + param.unit + '" ' + wording.toBe + ' "' + valueFixed + param.unit + '"',
                                node: decl,
                                result,
                                ruleName,
                                word: decl.params
                            });
                            param.value = valueFixed.toString();
                        }
                    });
                    if (context.fix) {
                        decl.params = params.toString();
                    }
                });
            });
            (_b = options.properties) === null || _b === void 0 ? void 0 : _b.map(property => {
                root.walkDecls(property, decl => {
                    const nodes = parse(decl.value);
                    nodes.walkNumerics(node => {
                        if (options.units.includes(node.unit) && Number(node.value) % options.step > 0) {
                            const valueFixed = Math.round(Number(node.value) / options.step) * options.step;
                            utils.report({
                                message: wording.expected + ' "' + property + '" ' + wording.unit + ' "' + node.value + node.unit + '" ' + wording.toBe + ' "' + valueFixed + node.unit + '"',
                                node: decl,
                                result,
                                ruleName,
                                word: decl.value
                            });
                            node.value = valueFixed.toString();
                        }
                    });
                    if (context.fix) {
                        decl.value = nodes.toString();
                    }
                });
            });
        }
    };
}
var unitStepRule = createPlugin(ruleName, rule);

var index = [
    noDisableRule,
    noObsoleteRule,
    unitStepRule
];

export { index as default };
